File: Applications/CSharp/Trees/IncreasingOrderSearch.cs

public class TreeNode
{
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null)
    {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Solution
{
    TreeNode next;
    public TreeNode IncreasingBST(TreeNode root)
    {
        if (root == null)
        {
            return null;
        }
        var beginning = next = new TreeNode();

        InOrder(root);
        return beginning.right;
    }

    private void InOrder(TreeNode node)
    {
        if (node == null)
            return;

        InOrder(node.left);
        node.left = null;
        next.right = node;
        next = node;
        InOrder(node.right);
    }
}

File: Applications/Cpp/Trees/IncreasingOrderSearch.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

struct TreeNode *next;

struct TreeNode *increasingBST(struct TreeNode *root)
{
    if (root == 0)
        return 0;

    struct TreeNode *beginning = next = (struct TreeNode *)malloc(sizeof(struct TreeNode));
    inOrder(root);
    return beginning->right;
}

void inOrder(struct TreeNode *node)
{
    if (node == 0)
    {
        return 0;
    }

    inOrder(node->left);
    node->left = 0;
    next->right = node;
    next = node;
    inOrder(node->right);
}